Corrections regarding issue #12

gotofoo opened the pull request:
I made a few changes to make this package ready for HAP-NodeJS 0.5.4 since the new version includes a check to make sure the MAC address (username) of the bridge is valid. If the address is not valid the publish function will fail. So far the generated address of the macify function was in the form of XX:XX:XX:XX:XX:XX:XX:XX instead of the XX:XX:XX:XX:XX:XX which HAP-NodeJS expects. Probably it would be a good idea to add an option in the node configuration to allow the user to change the MAC address manually.

The other few changes just make sure that the redeploy issues described in the issue https://github.com/NRCHKB/node-red-contrib-homekit-bridged/issues/12 should be solved now, at least for the most part. For me it is working now no matter if I deploy only modified nodes or if I perform a full deploy.

**It should be noted that all paired devices will need to be re-paired since the MAC address will change due to the fix in the macify function.**

crxporter commented:
Hey @shaquu can we get this in a dev branch with 1.0? I’m sure many of us will want to get testing and using this fix ASAP! 

Awesome work, @gotofoo!

Shaquu commented:
Yes of course. I just landed in my hometown. Give me a day and I should look at that.

radokristof commented:
If this is merged into dev I will also update my stable system to dev. I have some problems right now, I want to test the new version...

crxporter commented:
Also it’s been a while since I installed anything but the release branch - a quick reminder of how to install dev branch would be great!

Shaquu commented:
@gotofoo thanks for PR.

Regarding Mac address issue. It was already resolved on dev branch #170 #172 
Please mark if it works incorrectly on dev branch.

Regarding update to latest hap-nodejs (0.5.4).
I added it on dev branch as it is already in different PR (by greenkeeper).

If your PR makes more useful changes then please follow these steps:

- rebase into dev branch
- comment on changes

gotofoo commented:
@Shaquu I rebased the changes on the dev branch. The changes now only contain corrections regarding the issue #12. Here is a short description of all the changes:

- the isParentNode property is now stored in the node for later use in the onClose function, this was a bug since the onClose uses the isParentNode property even though it was only defined as a local constant in HAPServiceNode.js.
- all listerner functions are now stored in the node object to be able to remove them later if the node gets restarted
- all removeAllListerners in the onClose functions were replaced with removeListener because removeAllListerners could potentially mess up internal listeners of HAP-NodeJS
- the getOrCreate function was changed to support changes of the service type by removing the old service and replacing it with a new one
- removed the third parameter (accessory category) from "new Accessory(...)" since HAP-NodeJS only expects 2 parameters

I noticed that the checks are currently failing due to changes in the api.js file where the accessory category should automatically be determined based on the service type but not for all service types a default category was defined.

From my point of view the whole logic regarding accessory categories should be removed since it is not used for anything anyway. The only time when HAP-NodeJS uses the category is while publishing an accessory which makes sense since according to the HomeKit specification only the primary accessory you pair with can have a category. This means that in a bridged scenario only the bridge itself will have a category but not the bridged accessories within that bridge.

In the case of this repository the only category which is really needed is the category "BRIDGE" while publishing the bridge since all other accessories are just bridged. This would make configuration easier since the category would not need to be specified in the frontend.

Shaquu commented:
About accessory type/subtype. If you are sure about it you can remove it from current code as it might be useless indeed.

Shaquu commented:
@gotofoo will you find a time to do the change?

gotofoo commented:
@Shaquu The last couple of days I was a bit busy with a few other things. I will get the changes you requested done by the end of this week.

I will create a separate pull request for removing the accessory category logic. As I mentioned bridged accessories do not have a category so it does not make sense to specify a category. This will also help those failing checks ;)

gotofoo commented:
@Shaquu I made those changes you requested and testet a bit around. I was able to add and remove accessories/services and all the updates got pushed to the clients just fine.

Since the Homekit specification says that the values of the characteristics Name, Manufacturer, Serial Number and Model must not change during the lifetime of an accessory I changed the code a bit. Now the accessory will automatically be replaced with a new one if any of those values change. This causes the clients to remove the existing accessory and adding a new one. The UUID of the accessory is now generated based on those fields so that the UUID changes if any of those fields change which tells HAP-NodeJs to use a new accessory ID. It would probably be a good idea to add a warning message after any of those fields were changed to inform the user that this will cause the existing accessory to be replaced.

Furthermore I also included the change I mentioned regarding the accessory category which is not used. For this I removed the category field and all the code related to it including the code which caused the checks to fail.

Also I removed the fields Manufacturer, Serial Number and Model from linked service Nodes because those fields are not required for linked services.

Shaquu commented:
I am not sure about removing info fields from linked service as I don't remember how they works right now (maybe someone is using them for something?). Anyway since it's 1.0.0.dev I am merging.
